### PR TITLE
If field is a simple field return a simple resolver

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

This PR improves Strawberry performance by returning a simple resolver for fields without a custom resolver or any permission classes.

Benchmarks:

strawberry-simple-resolvers

Note: "simple" fields won't call the `get_result` function on the field. I've implemented the `is_simple_field` check as a property on the field so that custom fields can override it if they always need to post process results from the resolver.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Jonathan Kim for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/field.py [hunk 1]
```diff
@@ -287,6 +291,18 @@ def _has_async_base_resolver(self) -> bool:
...
 
+    @property
+    def is_simple_field(self):
+        """
+        Flag indicating if this is a "simple" field that just returns the
+        relevant attribute from the source object. If it is a simple field we
+        can avoid constructing an `Info` object and running any permission
+        checks in the resolver which improves performance.
+
+        Note: if a field is "simple" then `get_result` is not called.
```
Comment (patrick91) [inline strawberry/field.py:302]: does this mean I can't do:

class ACustomField(...):
   def get_result(...):
     return '123'

```
@strawberry.type
class User:
    something: str = ACustomField()
```

Comment (jkimbo) [reply strawberry/field.py:302]: You would need to override the property. Like this:

```python
class ACustomField(...):
	@property
	def is_simple_field(self):
		return False

	def get_result(...):
    	return '123'

@strawberry.type
class User:
    something: str = ACustomField()
```

File: strawberry/field.py [hunk 2]
```diff
@@ -34,6 +34,10 @@
 _RESOLVER_TYPE = Union[StrawberryResolver, Callable]
 
 
+def default_resolver(name: str, source: Any, *args, **kwargs):
+    return getattr(source, name)
```
Comment (BryceBeagle) [inline strawberry/field.py:38]: Can this be `StrawberryResolver`? Potentially constructed using a `classmethod`?

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: strawberry/field.py (hunk 2 above)
Comment (jkimbo) [reply strawberry/field.py:38]: I guess it can be though what would the benefit be?

Comment (BryceBeagle) [reply strawberry/field.py:38]: So that `Field.get_result` always deals with the same datatype

Comment (jkimbo) [reply strawberry/field.py:38]: Ok I see. Will change.

File: strawberry/schema/schema_converter.py [hunk 3]
```diff
@@ -53,6 +55,9 @@
 from .types.concrete_type import ConcreteType
 
 
+default_resolver = StrawberryResolver.default_resolver()
```
Comment (BryceBeagle) [inline strawberry/schema/schema_converter.py:58]: This is defined twice, here and in `field.py`.

File: strawberry/types/fields/resolver.py [hunk 4]
```diff
@@ -18,6 +18,10 @@
 T = TypeVar("T")
 
 
+def default_resolver(name: str, source: Any, *args, **kwargs):
```
Comment (BryceBeagle) [inline strawberry/types/fields/resolver.py:21]: I think this should live as an attribute on `StrawberryResolver`, probably private

File: strawberry/types/fields/resolver.py [hunk 5]
```diff
@@ -38,6 +42,10 @@ def __init__(
     def __call__(self, *args, **kwargs) -> T:
         return self.wrapped_func(*args, **kwargs)
 
+    @classmethod
+    def default_resolver(cls):
```
Comment (BryceBeagle) [inline strawberry/types/fields/resolver.py:46]:
```suggestion
    def default_resolver(cls) -> StrawberryResolver:
```

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: strawberry/types/fields/resolver.py (hunk 5 above)
Comment (jkimbo) [reply strawberry/types/fields/resolver.py:46]: Is mypy not clever enough to figure that out since I’m returning `cls`?

Comment (BryceBeagle) [reply strawberry/types/fields/resolver.py:46]: Maybe it is, but I tend towards being explicit

**Comment by jkimbo** [comment]
Because this PR is now quite out of date I've recreated it here: #2194

**Pull request closed without merging**